Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,312 @@
+            _actionsSystem.SetCharges(spell, charges < 0 ? null : charges);
+            component.Spells.Add(spell);
+        }
+    }
+
+    private void OnUse(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        AttemptLearn(uid, component, args);
+
+        args.Handled = true;
+    }
+
+    private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        component.CancelToken = new();
+        var doAfterEventArgs = new DoAfterEventArgs(args.User, component.LearnTime, component.CancelToken.Token, uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            BreakOnDamage = true,
+            BreakOnStun = true,
+            NeedHand = true, //What, are you going to read with your eyes only??
+            TargetFinishedEvent = new LearnDoAfterComplete(args.User),
+            TargetCancelledEvent = new LearnDoAfterCancel(),
+        };
+
+        _doAfter.DoAfter(doAfterEventArgs);
+    }
+
+    private void TeachSpell(EntityUid uid, SpellbookComponent component, LearnDoAfterComplete ev)
+    {
+        component.CancelToken = null;
+
+        if (ev.User == null)
+            return;
+
+        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+    }
+
+    #region Spells
+
+    private void OnRuneMagic(RuneMagicEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+        Spawn(args.RunePrototype, transform.Coordinates.SnapToGrid());
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Teleports the user to the clicked location
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnTeleportSpell(TeleportSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+
+        if (_mapManager.TryFindGridAt(args.Target, out var grid))
+        {
+            var gridPosition = grid.WorldToLocal(args.Target.Position);
+
+            transform.Coordinates = new EntityCoordinates(grid.GridEntityId, gridPosition);
+            SoundSystem.Play(Filter.Pvs(args.Target), args.BlinkSound.GetSound());
+        }
+        else
+        {
+            var mapEntity = _mapManager.GetMapEntityIdOrThrow(args.Target.MapId);
+            transform.AttachParent(mapEntity);
+            transform.WorldPosition = args.Target.Position;
+            SoundSystem.Play(Filter.Pvs(args.Target), args.BlinkSound.GetSound());
+        }
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Spawns 3 walls in front of the caster in a 3x1/1x3 pattern
+    /// Disappears after a set amount of time
+    /// Allows caster to walk through them freely.
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnForceWallSpell(ForceWallSpellEvent args)
+    {
+        if (args.Handled)
+            return;
```

Comment:
This should be configurable as right now i can't specify I want a line of walls or if I want 1 wall for example, I have to make an entirely new method to do it.

I also don't see anything that means rune / forcewall logic can't be combined; they both spawn anchored entities and the only diff is instantaction vs targeted when the underlying code could be the same between them.

You should generecise both of them, then add a method that takes in a position to spawn at (Instant supplying the player's pos and the targeted one being the target pos) which means you can combine both of their logics into one. Then you can look at making the spawn logic generic too.